hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/state/ReadableKVStateBase.java

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright (C) 2022 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.hedera.node.app.spi.state;
+
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.Nullable;
+import java.util.*;
+
+/**
+ * A base class for implementations of {@link ReadableKVState} and {@link WritableKVState}.
+ *
+ * @param <K> The key type
+ * @param <V> The value type
+ */
+public abstract class ReadableKVStateBase<K extends Comparable<K>, V>
+        implements ReadableKVState<K, V> {
+    /** The state key, which cannot be null */
+    private final String stateKey;
+
+    /**
+     * A cache of all values read from this {@link ReadableKVState}. If the same value is read
+     * twice, rather than going to the underlying merkle data structures to read the data a second
+     * time, we simply return it from the cache. We also keep track of all keys read, which is
+     * critical for dealing with validating what we read during pre-handle with what may have
+     * changed before we got to handle transaction. If the value is "null", this means it was NOT
+     * FOUND when we looked it up.
+     */
+    private final Map<K, V> readCache = new HashMap<>();
+
+    private final Set<K> unmodifiableReadKeys = Collections.unmodifiableSet(readCache.keySet());
+
+    /**
+     * Create a new StateBase.
+     *
+     * @param stateKey The state key. Cannot be null.
+     */
+    protected ReadableKVStateBase(@NonNull String stateKey) {
+        this.stateKey = Objects.requireNonNull(stateKey);
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    @NonNull
+    public final String getStateKey() {
+        return stateKey;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    @Nullable
+    public V get(@NonNull K key) {
+        // We need to cache the item because somebody may perform business logic basic on this
+        // contains call, even if they never need the value itself!
+        Objects.requireNonNull(key);
+        if (!hasBeenRead(key)) {
+            final var value = readFromDataSource(key);
+            markRead(key, value);
+        }
+        return readCache.get(key);
+    }
+
+    /**
+     * Gets the set of keys that a client read from the {@link ReadableKVState}.
+     *
+     * @return The possibly empty set of keys.
+     */
+    @NonNull
+    public final Set<K> readKeys() {
```

Comment:
I think this method needs to be defined in `ReadableKVState`.

This method not being available in the interface was probably the main reason why I missed that all implementations of `ReadableKVState` track their `readKeys`.

Reply:
Ah...